### PR TITLE
enhancement: improve json detection by detect_filetype

### LINKED ISSUE #704: partition_json and _is_text_file_a_json detect any JSON regardless of format
`partition_json` ([unstructured/partition/json.py](https://github.com/Unstructured-IO/unstructured/blob/559a5578ba5c798a35b9efb5852706c7cf99f2b7/unstructured/partition/json.py#L4)) right now is only meant for processing JSONs that are a dictionary representation of a list of Element objects, which is why it is checking for `[` at the beginning of the file (and also why it has the weird `LIST_OF_DICTS` name).

- `_is_text_file_a_json` and `partition_json` should detect any JSON regardless of format
- `partition_json` should get updated to do an actual check on the key values, etc. instead of just looking for a list of dicts.
- `partition_json` and `_is_text_file_a_json` both currently check for `LIST_OF_DICTS_PATTERN` and raises if appropriate (see unstructured/nlp/patterns.py)

### PR DESCRIPTION
**Pull request by Coniferish**
## Summary
Closes #704 
Updates `_is_text_file_a_json` to detect more general json format (not just a list of dicts)
Adds associated tests

## Testing
```
from unstructured.file_utils.filetype import detect_filetype
detect_filetype(filename = "example-docs/fake-incomplete-json.txt")
```

### CODE REVIEW

**Comment by Coniferish** [comment]
@MthwRobinson I overly commented the regex pattern so it's easy for you to see how I was thinking about the pattern. Let me know if you have any thoughts on that. Can remove the comments once you take a look.

File: unstructured/nlp/patterns.py [hunk 1]
```diff
@@ -108,4 +108,17 @@
...
+# (?s) dot all (including newline characters)
+# \{(?=.*:) opening brace and at least one colon
+# .*? any characters (non-greedy)
+# (?:\}|$) non-capturing group that matches either the closing brace } or the end of
+# the string to handle cases where the JSON is cut off
+# | or
+# \[(?s:.*?)\] matches the opening bracket [ in a JSON array and any characters inside the array
+# (?:$|,|\]) non-capturing group that matches either the end of the string, a comma,
+# or the closing bracket to handle cases where the JSON array is cut off
+JSON_PATTERN = r"(?s)\{(?=.*:).*?(?:\}|$)|\[(?s:.*?)\](?:$|,|\])"
```
Comment (MthwRobinson) [inline unstructured/nlp/patterns.py:121]: Nice!

**Review by MthwRobinson — APPROVED** [review]
LGTM! Nice regexing!

**Comment by MthwRobinson** [comment]
@Coniferish - You can leave the regex comments in there. That'll make it easier for future us to decipher ...

**Pull request merged**